tests/kernel/interrupt/src/dynamic_shared_irq.c

```diff
@@ -0,0 +1,280 @@
+		     "unexpected client data for irq1, index 1");
+
+	arch_irq_connect_dynamic(fixture.irq2, fixture.irq_priority,
+				 test_isr_2, (void *)2, 0);
+
+	zassert_true(_sw_isr_table[fixture.irq2_table_idx].isr == test_isr_2,
+		     "wrong _sw_isr_table ISR at irq2");
+	zassert_true(_sw_isr_table[fixture.irq2_table_idx].arg == (void *)2,
+		     "wrong _sw_isr_table argument at irq2");
+	zassert_true(!_shared_sw_isr_table[fixture.irq2_table_idx].client_num,
+		     "wrong client number at irq2");
+
+	reset_test_vector();
+}
+
+/**
+ * @brief Test writing to a vector with a shared interrupt
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details This tests if interrupts are dynamically shared successfully
+ * (i.e: multiple ISR/arg pairs are called whenever the interrupt
+ * they were registered for is triggered).
+ */
+ZTEST(shared_irq_feature, test_dynamic_shared_irq_write)
+{
+	int i;
+
+	irq_enable(fixture.irq1);
+	irq_enable(fixture.irq2);
+
+	trigger_irq(fixture.irq1);
+	trigger_irq(fixture.irq2);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		zassert_true(test_vector[i] == result_vector[i],
+			     "wrong test_vector value at %d: 0x%x vs 0x%x",
+			     i, test_vector[i], result_vector[i]);
+	}
+
+	irq_disable(fixture.irq1);
+	irq_disable(fixture.irq2);
+}
+
+/**
+ * @brief Test writing to a vector after an ISR/arg disconnect.
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details This tests if ISR/arg pairs are disconnected successfully
+ * and the interrupts are "unshared" whenever a single ISR/arg pair is
+ * left.
+ */
+ZTEST(shared_irq_feature, test_dynamic_shared_irq_disconnect_write)
+{
+	int i;
+
+	/* remove test_isr_0/NULL pair. After this statement we expect
+	 * irq1 to be unshared.
+	 */
+	arch_irq_disconnect_dynamic(fixture.irq1, fixture.irq_priority,
+				    test_isr_0, 0, 0);
+
+	zassert_true(_sw_isr_table[fixture.irq1_table_idx].isr == test_isr_1,
+		     "wrong _sw_isr_table ISR at irq1");
+	zassert_true(_sw_isr_table[fixture.irq1_table_idx].arg == (void *)1,
+		     "wrong _sw_isr_table arg at irq1");
+	zassert_true(!_shared_sw_isr_table[fixture.irq1_table_idx].client_num,
+		     "wrong client number at irq1");
+
+	irq_enable(fixture.irq1);
+	trigger_irq(fixture.irq1);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		if (i == 1) {
+			zassert_true(test_vector[i] == result_vector[i],
+				     "wrong test_vector at %d: 0x%x vs 0x%x",
+				     i, test_vector[i], result_vector[i]);
+			continue;
+		}
+
+		zassert_true(!test_vector[i],
+			     "wrong test_vector value at %d: 0x%x vs 0x%x",
+			     i, test_vector[i], result_vector[i]);
+	}
+
+	irq_disable(fixture.irq1);
+}
+
+ZTEST_SUITE(shared_irq_feature, NULL,
+	    dynamic_shared_irq_suite_setup,
+	    dynamic_shared_irq_suite_before,
+	    dynamic_shared_irq_suite_after,
```

Comment:
can you add at least for one architecture a test for the statically allocated shared IRQ lines?

Reply:
ACK. My main concern with using static shared interrupts is that the test becomes brittle. What I mean by this is that if the interrupt lines I choose are not usable on all platforms for said architecture then build errors will appear due to the number of clients exceeding the limit. Since we're limiting this to a single architecture then I guess it may not be so bad, but if more testing platforms (or even tests) are added then it may become problematic.

Reply:
yeah, but if you restrict the test to something well know like some QEMU platform or even `native_posix` you should not get into any trouble.